### PR TITLE
[SPARK-33723][SQL] ANSI mode: Casting String to Date should throw exception on parse error

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Currently, when casting a string as timestamp type in ANSI mode, Spark throws a runtime exception on parsing error.
However, the result for casting a string to date is always null. We should throw an exception on parsing error as well.

### Why are the changes needed?

Add missing feature for ANSI mode

### Does this PR introduce _any_ user-facing change?

Yes for ANSI mode, Casting string to date will throw an exception on parsing error

### How was this patch tested?

Unit test

### CODE REVIEW

File: docs/sql-ref-ansi-compliance.md [hunk 1]
```diff
@@ -163,6 +163,7 @@ The behavior of some SQL operators can be different under ANSI mode (`spark.sql.
   - `array_col[index]`: This operator throws `ArrayIndexOutOfBoundsException` if using invalid indices.
   - `map_col[key]`: This operator throws `NoSuchElementException` if key does not exist in map.
   - `CAST(string_col AS TIMESTAMP)`: This operator should fail with an exception if the input string can't be parsed.
+  - `CAST(string_col AS Date)`: This operator should fail with an exception if the input string can't be parsed.
```
Comment (HyukjinKwon) [inline docs/sql-ref-ansi-compliance.md:166]: one nit: `Date` -> `DATE`?

**Comment by HyukjinKwon** [comment]
cc @MaxGekk too

**Review by MaxGekk — COMMENTED** [review]
How about other places similar to CAST, for example:
https://github.com/apache/spark/blob/b112e2bfa619d028004cbc7fb8ec1363689729a7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1932

Should we take into account the ANSI flag there?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -466,6 +466,15 @@ object DateTimeUtils {
     }
   }
 
+  def stringToDateAnsi(s: UTF8String, zoneId: ZoneId): Int = {
+    val date = stringToDate(s, zoneId)
+    if (date.isEmpty) {
+      throw new DateTimeException(s"Cannot cast $s to DateType.")
+    } else {
+      date.get
+    }
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:475]:
```suggestion
    stringToDate(s, zoneId).getOrElse {
      throw new DateTimeException(s"Cannot cast $s to DateType.")
    }
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 3]
```diff
@@ -984,6 +984,30 @@ abstract class AnsiCastSuiteBase extends CastSuiteBase {
     }
   }
 
+  test("ANSI mode: cast string to date with parse error") {
+    val activeConf = conf
+    new ParVector(ALL_TIMEZONES.toVector).foreach { zid =>
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:989]: Do we really need to check all time zones? Would `DateTimeTestUtils.outstandingZoneIds` be enough?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:989]: +1

Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:989]: This is from the test case for timestamp. 
+1, let me update both.

**Comment by gengliangwang** [comment]
> How about other places similar to CAST

@MaxGekk there are many other places calling `stringToTimestamp` and `stringToDate` directly. We can have follow-ups to resolve them.

**Comment by gengliangwang** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 4]
```diff
@@ -466,6 +466,12 @@ object DateTimeUtils {
     }
   }
 
+  def stringToDateAnsi(s: UTF8String, zoneId: ZoneId): Int = {
+    stringToDate(s, zoneId).getOrElse {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:470]: can you also update `stringToTimestampAnsi` to follow this code style?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:467]: sure, updated.

**Comment by maropu** [comment]
lgtm

**Review by cloud-fan — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by gengliangwang** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**